fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -453,15 +453,47 @@ def __init__(
         to_be_flatten_params: List[List[Parameter]] = [[]]
         non_flatten_params = params
         param_name_groups = [[n] for n in param_names]
+
+        # This is the suffix list that we appended to flattened parameters to seperate parameters of weight, bias and the rest of
+        # the parameters.
+        param_names_suffix = []
         if self.flatten_parameters:
-            to_be_flatten_params = [params]
+            # We put the params to three groups. weight, bias and the rest of parameters
+            to_be_flatten_params = []
             non_flatten_params = []
-            param_name_groups = [param_names]
+            del param_name_groups
+            group_to_index = {}
+            param_name_groups = []
+
+            for i, n in enumerate(param_names):
+                pg = getattr(params[i], "param_group", "default")
```

Comment:
If I understood it correctly, the original module's param get tagged with an attr named "param_group", which has the str type, right? Then, FSDP uses this tag to put params in groups and all untagged ones fall into the default flatten group?

If so, how do user express a need to leave params unflattened? Also, is the tagging process easy and not error-prune for the users / caller code? Can you point to me where is the caller code for this in your branch?

I hope we can explore a few options for this API to find one that is easiest and less error prone yet most flexible for the users. :-) Thanks for this initial idea though. I do think it could be the one we end up using. Just wanted to carefully consider a few alternatives.

Reply:
the composite optimizer in fairseq works by groupping params into different groups based on this "param_group" attribute. if its absent, the param gets put into a "default" group.

keeping params unflattened i guess would be accomplished by setting "flatten_parameters" to false, right ?

the tagging process boils down to just setting props on params in model constructor. e.g. see: https://github.com/facebookresearch/fairseq/blob/main/examples/wav2vec/unsupervised/models/wav2vec_u.py#L458

Reply:
I see. Will read in more details.
 
> keeping params unflattened i guess would be accomplished by setting "flatten_parameters" to false, right ?
I see, it is a combination of the tag and the arg flag. Then, if user wants to completely disable flattening, he/she needs to BOTH remove the tags (p.param_group) and set the arg flag to False? That could be a bit error-prune because I can see some users just flip and flag and forgot to remove the tags. ;-)